ember-resources/src/util/function.ts

```diff
@@ -110,25 +113,43 @@ export function trackedFunction<Return>(...passedArgs: UseFunctionArgs<Return>)
         state.error = e;
       } finally {
         state.isResolved = true;
+        console.log("GV Finished");
       }
-    })();
-
-    return state;
+    };
+
+    const trackedFunctionProperty = new TrackedFunctionProperty(
+      getValue,
+      initialValue
+    );
+    getValue(trackedFunctionProperty.state);
+    return trackedFunctionProperty;
   });
 }
 
-/**
- * State container that represents the asynchrony of a `trackedFunction`
- */
-export class State<Value> {
-  @tracked isResolved = false;
-  @tracked resolvedValue?: Value;
-  @tracked error?: unknown;
+export class TrackedFunctionProperty<Value> {
+  @tracked state: State<Value>;
 
-  constructor(public initialValue?: Value) {}
+  constructor(
+    private getValue: (state: State<Value>) => {},
+    initialValue?: Value
+  ) {
+    this.state = new State(initialValue);
+  }
+
+  execute() {
+    this.state = new State(this.state.initialValue);
+    this.getValue(this.state);
+    return this.state;
```

Comment:
Do you know a way around this?
I think this is the line that causes the values to just no update on reassignment:
https://github.com/NullVoxPopuli/ember-resources/blob/main/ember-resources/src/core/function-based/utils.ts#L29

Reply:
yeah, so, it's a little silly, but in the original code, if you were to do:
```ts
  return resource<State<Return>>(context, (hooks) => {
    let state = new State(initialValue, fn, hooks);

    (async () => state.retry())();

    // retry will update the tracked data on State appropriately
    return state;
// ...

class State {
  constructor(public initialValue?: Value, private fn: ResourceFn<Value>, private hooks: Hooks) {}
  
  async retry() {
      try {
        let notQuiteValue = this.fn(this.hooks);
        let promise = Promise.resolve(notQuiteValue);

        waitForPromise(promise);

        let result = await promise;

        this.error = undefined;
        this.resolvedValue = result;
      } catch (e) {
        this.error = e;
      } finally {
        this.isResolved = true;
      }

  }
}
```

Reply:
Just to be clear since im not sure i explained well enough what the problem was, my problem is with resetting for example `isResolved` from `true` back to `false` while `retry` is running. The values are updated correctly after `retry` is run, i just cant reset them while `retry` is running.

Reply:
> i just cant reset them

can you expand on this more?

Maybe starting with a unit test to describe the behavior and order of that behavior would be helpful?

those are here: https://github.com/NullVoxPopuli/ember-resources/blob/main/testing/ember-app/tests/utils/function/rendering-test.ts

Reply:
how'd it go?

Reply:
Im sorry, i was on holidays the last few weeks. Im back since today, i think i will have some time to look into it.

Reply:
Im sorry, i was on holidays the last few weeks. Im back since today, i think i will have some time to look into it.

Reply:
all good! no worries